src/Microsoft.AspNet.Authentication/RemoteAuthenticationHandler.cs

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http.Features.Authentication;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Authentication
+{
+    public abstract class RemoteAuthenticationHandler<TOptions> : AuthenticationHandler<TOptions> where TOptions : RemoteAuthenticationOptions
+    {
+        public override async Task<bool> HandleRequestAsync()
+        {
+            if (Options.CallbackPath.HasValue && Options.CallbackPath == Request.Path)
+            {
+                return await HandleRemoteCallbackAsync();
+            }
+            return false;
+        }
+
+        protected virtual async Task<bool> HandleRemoteCallbackAsync()
+        {
+            var authResult = await HandleRemoteAuthenticateAsync();
+            if (authResult == null || !authResult.Succeeded)
+            {
+                var errorContext = new ErrorContext(Context, authResult?.Error ?? new Exception("Invalid return state, unable to redirect."));
+                Logger.LogInformation("Error from RemoteAuthentication: " + errorContext.Error.Message);
+                await Options.Events.RemoteError(errorContext);
+                if (errorContext.HandledResponse)
+                {
+                    return true;
+                }
+                if (errorContext.Skipped)
+                {
+                    return false;
+                }
+
+                Context.Response.StatusCode = 500;
+                return true;
+            }
+
+            // We have a ticket if we get here
+            var ticket = authResult.Ticket;
+            var context = new TicketReceivedContext(Context, ticket)
```

Comment:
This prevents them from changing what sign in scheme to use if we did this.  Although I'm not sure if that's really a scenario we want to support